jax/_src/config.py

```diff
@@ -263,27 +263,56 @@ def get_state(self):
       return val if val is not unset else self._read(name)
     setattr(Config, name, property(get_state))
 
-    @contextlib.contextmanager
-    def set_state(new_val: Optional[str]):
+    def validate(new_val):
       if (new_val is not None and
           (type(new_val) is not str or new_val not in enum_values)):
         raise ValueError(f"new enum value must be None or in {enum_values}, "
                          f"got {new_val} of type {type(new_val)}.")
-      prev_val = getattr(_thread_local_state, name, unset)
-      setattr(_thread_local_state, name, new_val)
-      if update_thread_local_hook: update_thread_local_hook(new_val)
-      try:
-        yield
-      finally:
-        if prev_val is unset:
-          delattr(_thread_local_state, name)
-          if update_thread_local_hook: update_thread_local_hook(None)
-        else:
-          setattr(_thread_local_state, name, prev_val)
-          if update_thread_local_hook: update_thread_local_hook(prev_val)
-    set_state.__name__ = name[4:] if name.startswith('jax_') else name
-    set_state.__doc__ = f"Context manager for `{name}` config option.\n\n{help}"
-    return set_state
+
+    return _StateContextManager(name, help, update_thread_local_hook, validate)
+
+  def define_string_state(
+      self, name: str, default: Optional[str], help: str,
+      update_global_hook: Optional[Callable[[str], None]] = None,
+      update_thread_local_hook: Optional[Callable[[Optional[str]], None]]
+      = None):
```

Comment:
```suggestion
      = None) -> ContextManager[None]:
```

(I _think_ that annotation works...)

And maybe join the `= None` part onto the previous line, followed by a newline? I dno. It looks lonely.

Reply:
Seems to work! I put `= None` with its friends.

Reply:
Actually the ContextManager annotation doesn't work (and I apparently don't know how to run mypy). It says:
`Incompatible return value type (got "_StateContextManager", expected "ContextManager[None]")`
I'm guessing because we're using a weird context manager pattern. I'm just gonna omit it, unless you have ideas.